File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.cc
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef pair < ll, ll>pairll;

const ll maxN = 300004;

vector<ll>vec[maxN];
ll N, M, K, X, num1, num2;
ll visited[maxN];


int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0); 
	cout.tie(0);

	cin >> N >> M >> K >> X;

	// 배열 생성
	while (M--) {
		cin >> num1 >> num2;
		vec[num1].push_back(num2);
	}

	vector<ll> resultVec;
	queue<ll>que;
	que.push(X);
	visited[X] = 1;

	while (que.size()) {
		ll queSize = que.size();
		for (ll i = 0; i < queSize; i++) {
			ll current = que.front(); que.pop();

			for (ll n : vec[current]) {
				if (visited[n] != 0) continue;
				visited[n] = visited[current] + 1;
				if (visited[n] - 1 == K) resultVec.push_back(n);
				que.push(n);

			}
		}
	}

	sort(resultVec.begin(), resultVec.end());
	if (resultVec.size() == 0) {
		cout << -1;
		return 0;
	}
	string result = "";
	for (ll r : resultVec) {
		result += (to_string(r) + "\n");
	}
	cout << result;
	
	return 0;
}

File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.java
import java.util.*;

public class Main {
    static final int MAXN = 300004;
    static List<Integer>[] vec = new ArrayList[MAXN];
    static int[] visited = new int[MAXN];

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        int M = sc.nextInt();
        int K = sc.nextInt();
        int X = sc.nextInt();

        for (int i = 0; i < N + 1; i++) {
            vec[i] = new ArrayList<>();
        }

        for (int i = 0; i < M; i++) {
            int num1 = sc.nextInt();
            int num2 = sc.nextInt();
            vec[num1].add(num2);
        }

        List<Integer> resultVec = new ArrayList<>();
        Queue<Integer> que = new LinkedList<>();
        que.add(X);
        visited[X] = 1;

        while (!que.isEmpty()) {
            int queSize = que.size();
            for (int i = 0; i < queSize; i++) {
                int current = que.poll();

                for (int n : vec[current]) {
                    if (visited[n] != 0) continue;
                    visited[n] = visited[current] + 1;
                    if (visited[n] - 1 == K) resultVec.add(n);
                    que.add(n);
                }
            }
        }

        Collections.sort(resultVec);
        if (resultVec.isEmpty()) {
            System.out.println(-1);
        } else {
            StringBuilder result = new StringBuilder();
            for (int r : resultVec) {
                result.append(r).append("\n");
            }
            System.out.print(result);
        }

        sc.close();
    }
}